src/host.rs

```diff
@@ -88,6 +133,151 @@ impl Udp {
     }
 
     pub(crate) fn unbind(&mut self, addr: SocketAddr) {
-        assert!(self.binds.remove(&addr).is_some(), "unknown bind {}", addr)
+        assert!(self.binds.remove(&addr).is_some(), "unknown bind {}", addr);
+
+        trace!(?addr, protocol = %"UDP", "Unbind");
+    }
+}
+
+pub(crate) struct Tcp {
+    /// Bound server sockets
+    binds: IndexMap<SocketAddr, mpsc::UnboundedSender<(Syn, SocketAddr)>>,
+
+    /// Active stream sockets
+    sockets: IndexMap<SocketPair, StreamSocket>,
+}
+
+struct StreamSocket {
+    buf: IndexMap<u64, StreamData>,
+    next_send_seq: u64,
+    recv_seq: u64,
+    sender: mpsc::UnboundedSender<StreamData>,
+}
+
+impl StreamSocket {
+    fn new() -> (Self, mpsc::UnboundedReceiver<StreamData>) {
+        let (tx, rx) = mpsc::unbounded_channel();
+        let sock = Self {
+            buf: IndexMap::new(),
+            next_send_seq: 1,
+            recv_seq: 0,
+            sender: tx,
+        };
+
+        (sock, rx)
+    }
+
+    fn assign_seq(&mut self) -> u64 {
+        let seq = self.next_send_seq;
+        self.next_send_seq += 1;
+        seq
+    }
+
+    // Buffer and re-order received segments by `seq` as the network may deliver
+    // them out of order.
+    fn buffer(&mut self, seq: u64, data: StreamData) {
+        assert!(
+            self.buf.insert(seq, data).is_none(),
+            "duplicate segment {}",
+            seq
+        );
+
+        while self.buf.contains_key(&(self.recv_seq + 1)) {
+            self.recv_seq += 1;
+
+            let data = self.buf.remove(&self.recv_seq).unwrap();
+            let _ = self.sender.send(data);
+        }
+    }
+}
+
+impl Tcp {
+    fn install() -> Self {
```

Comment:
why not "new". Are you trying to signify it's a. singleton per host?

Reply:
The _host has L4_ software really resonated with me :) I can use new instead if this is overboard.